### PR TITLE
Expose the status code to the caller of client

### PR DESCRIPTION
**Pull request by PhilipGough**
In the istio adapter, we want to gather response codes from 3scale and report them to metrics backend. This change allows the user of the client to obtain the status code as part of a custom error function

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -117,6 +128,6 @@ func handleErrResp(resp *http.Response) string {
 }
 
 // Helper method to generate error message for client functions
-func genRespErr(ep string, err string) error {
-	return fmt.Errorf("error calling %s endpoint - %s", ep, err)
+func genRespErr(code int, ep, err string) error {
```
Comment (unleashed) [inline client/client.go]: small nit here - can we use something like this? (pseudo-code):
```go
func genRespErr(resp *http.Response, ep, err string) error {
  if err == "" { // or maybe we can use err == nil?
    err := handleErrResp(resp)
  }

  return ApiErr{resp.StatusCode, ep, error}
}
```

**Comment by PhilipGough** [comment]
@unleashed yes totally agree. Lots of improvements to be made in this repo. I'll address your comment above

File: client/app.go [hunk 2]
```diff
@@ -32,12 +33,43 @@ func (c *ThreeScaleClient) CreateApp(accessToken string, accountId string, planI
...
+			code: resp.StatusCode,
+			err:  msg,
+		}
+	}
+	return nil
+}
+
+// Decodes and transforms an API response error into a string
+func handleErrResp2(resp *http.Response, expectCode int) error {
+	if resp.StatusCode != expectCode {
```
Comment (unleashed) [inline client/app.go]: this func does nothing when the status code is the expected one, so how about only calling it when we know for sure an error will be returned?

File: client/app.go [hunk 3]
```diff
@@ -32,12 +33,43 @@ func (c *ThreeScaleClient) CreateApp(accessToken string, accountId string, planI
...
+		return ApiErr{
+			code: resp.StatusCode,
+			err:  msg,
+		}
+	}
+	return nil
+}
+
+// Decodes and transforms an API response error into a string
+func handleErrResp2(resp *http.Response, expectCode int) error {
```
Comment (unleashed) [inline client/app.go]: will need to rename this! :)

Comment (PhilipGough) [reply client/app.go]: For sure. It clashes with existing. When you are happy this is in a good place I will begin to rewrite those other functions

File: client/app.go (hunk 2 above)
Comment (PhilipGough) [reply client/app.go]: Yep, makes sense

File: client/client.go (hunk 1 above)
Comment (PhilipGough) [reply client/client.go]: @unleashed I've pushed some changes here

**Comment by PhilipGough** [comment]
@unleashed sure yeah the placeholder naming was, well .. a placeholder :) will do all of the above. Thanks

**Pull request merged**